File: 09/solver.py
import numpy as np
from pathlib import Path
import itertools


def solve_first(file: Path):
    with open(file) as fh:
        lines = fh.readlines()
    lines = [line.strip() for line in lines]

    total = 0
    for line in lines:
        numbers = [int(x) for x in line.split() if x]
        all_differences = [
            numbers,
        ]
        differences = [
            second - first for (first, second) in itertools.pairwise(numbers)
        ]
        while not all([d == 0 for d in differences]):
            all_differences.append(differences)
            differences = [
                second - first for (first, second) in itertools.pairwise(differences)
            ]

        for i in reversed(range(len(all_differences) - 1)):
            all_differences[i].append(
                all_differences[i][-1] + all_differences[i + 1][-1]
            )

        res = all_differences[0][-1]
        total += res

    print("A:", total)


def solve_second(file: Path):
    with open(file) as fh:
        lines = fh.readlines()
    lines = [line.strip() for line in lines]

    total = 0
    for line in lines:
        numbers = [int(x) for x in line.split() if x]
        all_differences = [
            numbers,
        ]
        differences = [
            second - first for (first, second) in itertools.pairwise(numbers)
        ]
        while not all([d == 0 for d in differences]):
            all_differences.append(differences)
            differences = [
                second - first for (first, second) in itertools.pairwise(differences)
            ]

        for i in reversed(range(len(all_differences) - 1)):
            all_differences[i].insert(
                0,
                all_differences[i][0] - all_differences[i + 1][0],
            )

        res = all_differences[0][0]
        total += res

    print("B:", total)


def solve(file: Path):
    solve_first(file)
    solve_second(file)


if __name__ == "__main__":
    import fire

    fire.Fire(solve)

File: 09/solver.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// Return the pairwise difference of elements in a slice.
func pairwiseDifference(nums []int) []int {
	var diffs []int
	for i := 0; i < len(nums)-1; i++ {
		diffs = append(diffs, nums[i+1]-nums[i])
	}
	return diffs
}

// Check if all elements in the slice are zeros.
func allZeros(nums []int) bool {
	for _, num := range nums {
		if num != 0 {
			return false
		}
	}
	return true
}

func solveFirst(lines []string) int {
	total := 0
	for _, line := range lines {
		parts := strings.Fields(line)
		numbers := make([]int, len(parts))
		for i, part := range parts {
			numbers[i], _ = strconv.Atoi(part)
		}

		allDifferences := [][]int{numbers}
		differences := pairwiseDifference(numbers)
		for !allZeros(differences) {
			allDifferences = append(allDifferences, differences)
			differences = pairwiseDifference(differences)
		}

		for i := len(allDifferences) - 2; i >= 0; i-- {
			last := allDifferences[i][len(allDifferences[i])-1]
			nextLast := allDifferences[i+1][len(allDifferences[i+1])-1]
			allDifferences[i] = append(allDifferences[i], last+nextLast)
		}

		res := allDifferences[0][len(allDifferences[0])-1]
		total += res
	}
	return total
}

func solveSecond(lines []string) int {
	total := 0
	for _, line := range lines {
		parts := strings.Fields(line)
		numbers := make([]int, len(parts))
		for i, part := range parts {
			numbers[i], _ = strconv.Atoi(part)
		}

		allDifferences := [][]int{numbers}
		differences := pairwiseDifference(numbers)
		for !allZeros(differences) {
			allDifferences = append(allDifferences, differences)
			differences = pairwiseDifference(differences)
		}

		for i := len(allDifferences) - 2; i >= 0; i-- {
			first := allDifferences[i][0]
			nextFirst := allDifferences[i+1][0]
			allDifferences[i] = append([]int{first - nextFirst}, allDifferences[i]...)
		}

		res := allDifferences[0][0]
		total += res
	}
	return total
}

func solve(filepath string) {
	file, err := os.Open(filepath)
	if err != nil {
		fmt.Println("Error opening file:", err)
		return
	}
	defer file.Close()

	var lines []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}

	fmt.Println("A:", solveFirst(lines))
	fmt.Println("B:", solveSecond(lines))
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Please provide a file path")
		return
	}
	solve(os.Args[1])
}
